### PR TITLE
Add context menu entry to stop music

### LINKED ISSUE #294: Double clicking a song category stops music AND collapses/expands it.
Just a bit janky. Should probably just stop the music since there's already a button immediately to the left that does the same thing, and can be annoying when just trying to stop the music as you may need to then expand the category again.

### PR DESCRIPTION
**Pull request by in1tiate**
Also included in this commit are:
- Changes to the music packet handler that change the text from "has played a song" to "has stopped the music" in the case that the dummy song "STOPSONG" is played
- A condition stopping music categories from triggering a music packet when double-clicked

This should fully resolve #294.

Note that this functionality depends on support serverside -the companion PR is [tsuserver3/pull/119](https://github.com/AttorneyOnline/tsuserver3/pull/119).

### CODE REVIEW

**Comment by oldmud0** [comment]
What if instead of sending STOPSONG, we simply send an empty string?

**Comment by in1tiate** [comment]
> 
> 
> What if instead of sending STOPSONG, we simply send an empty string as the track title?

That would also solve the issue of being able to manually create a track named "STOPSONG" and have it play instead of properly stopping the music player. I'll go ahead and implement that.

**Comment by oldmud0** [comment]
TODO: update protocol docs

File: src/courtroom.cpp [hunk 1]
```diff
@@ -2604,8 +2604,16 @@ void Courtroom::append_ic_text(QString p_text, QString p_name, QString p_action,
 
   // Format the name of the actor
   ui_ic_chatlog->textCursor().insertText(p_name, bold);
+  // Special case for stopping the music
+  if (p_action == "has stopped the music") {
```
Comment (oldmud0) [inline src/courtroom.cpp:2608]:
```suggestion
  if (p_action == tr("has stopped the music")) {
```
Needs `tr()` - otherwise won't work as intended.

**Review by oldmud0 — DISMISSED** [review]
I don't like the way one specifies an "action" to use a specific sentence structure, but that's not something that can be fixed here.

File: src/courtroom.cpp [hunk 2]
```diff
@@ -2604,8 +2604,16 @@ void Courtroom::append_ic_text(QString p_text, QString p_name, QString p_action,
...
   // Format the name of the actor
   ui_ic_chatlog->textCursor().insertText(p_name, bold);
+  // Special case for stopping the music
+  if (p_action == tr("has stopped the music")) {
+    ui_ic_chatlog->textCursor().insertText(" " + p_action, normal);
+    if (log_newline)
+      ui_ic_chatlog->textCursor().insertHtml("<br>");
+    // Add a period because grammar is cool I guess
+    ui_ic_chatlog->textCursor().insertText(p_text + ".", normal);
+  }
```
Comment (Crystalwarrior) [inline src/courtroom.cpp:2614]: Ughhhhhhhhhh I really hate this

File: src/courtroom.cpp [hunk 3]
```diff
@@ -4076,12 +4093,14 @@ void Courtroom::on_music_list_double_clicked(QTreeWidgetItem *p_item,
   if (ao_app->effects_enabled)
     packet_contents.append(QString::number(music_flags));
   ao_app->send_server_packet(new AOPacket("MC", packet_contents), false);
+  qDebug() << packet_contents;
```
Comment (Crystalwarrior) [inline src/courtroom.cpp:4096]: You left behind a qDebug()

File: src/courtroom.cpp [hunk 4]
```diff
@@ -4076,12 +4093,14 @@ void Courtroom::on_music_list_double_clicked(QTreeWidgetItem *p_item,
...
 }
 
 void Courtroom::on_music_list_context_menu_requested(const QPoint &pos)
 {
   QMenu *menu = new QMenu();
-
+  menu->addAction(QString(tr("Stop Current Song")), this, SLOT(music_stop()));
+  menu->addSeparator();
```
Comment (Crystalwarrior) [inline src/courtroom.cpp:4103]: me likey, but why the seperator?

File: src/courtroom.cpp [hunk 5]
```diff
@@ -4159,6 +4178,21 @@ void Courtroom::music_list_collapse_all()
...
+  QStringList packet_contents; //its music list
+  packet_contents.append(""); // this is our fake song, playing it triggers special code
+  packet_contents.append(QString::number(m_cid));
+  if ((!ui_ic_chat_name->text().isEmpty() && ao_app->cccc_ic_support_enabled) ||
+      ao_app->effects_enabled)
+    packet_contents.append(ui_ic_chat_name->text());
+  if (ao_app->effects_enabled)
+    packet_contents.append(QString::number(music_flags));
+  ao_app->send_server_packet(new AOPacket("MC", packet_contents), false);
+  qDebug() << packet_contents;
```
Comment (Crystalwarrior) [inline src/courtroom.cpp:4193]: Forgot a debug

**Review by Crystalwarrior — CHANGES_REQUESTED** [review]

File: src/courtroom.cpp (hunk 4 above)
Comment (in1tiate) [reply src/courtroom.cpp:4102]: No real particular reason, honestly

**Comment by in1tiate** [comment]
Now also resolves #304

File: src/courtroom.cpp (hunk 4 above)
Comment (in1tiate) [reply src/courtroom.cpp]: Removed

File: src/courtroom.cpp [hunk 6]
```diff
@@ -4159,6 +4188,22 @@ void Courtroom::music_list_collapse_all()
   ui_music_list->setCurrentItem(current);
 }
 
+void Courtroom::music_stop()
+{               // send a fake music packet with a nonexistent song
+  if (is_muted) // this requires a special exception for "~stop.mp3" in
+    return;     // tsuserver3, as it will otherwise reject songs not on
+  QStringList packet_contents; // its music list
+  packet_contents.append(
+      "~stop.mp3"); // this is our fake song, playing it triggers special code
```
Comment (Crystalwarrior) [inline src/courtroom.cpp:4197]: Vanilla support, backwards compatibility, cool and good

File: src/courtroom.cpp [hunk 7]
```diff
@@ -4063,10 +4092,10 @@ void Courtroom::on_music_list_double_clicked(QTreeWidgetItem *p_item,
 {
   if (is_muted)
     return;
-
+  if (p_item->parent() == nullptr) // i.e. we've clicked a category
+    return;
```
Comment (Crystalwarrior) [inline src/courtroom.cpp:4096]: Wait is this a regression of double-clicking on music category stopping the music functionality? Shouldn't we make a settings option for this? I expect lots of people who are used to the old inferior style to complain they have no option to bring back old "ux is my passion" way.

File: src/courtroom.cpp [hunk 8]
```diff
@@ -3885,7 +3914,7 @@ void Courtroom::set_effects_dropdown()
     return;
   }
 
-  effectslist.prepend("None");
+  effectslist.prepend(tr("None"));
```
Comment (Crystalwarrior) [inline src/courtroom.cpp:3917]: oops lol

File: src/courtroom.cpp [hunk 9]
```diff
@@ -3135,14 +3155,15 @@ void Courtroom::handle_song(QStringList *p_contents)
...
-    if (channel == 0)
+    if (f_song == "~stop.mp3")
+      ui_music_name->setText(tr("None"));
+    else if (channel == 0)
       ui_music_name->setText(f_song_clear);
   }
   else {
     QString str_char = char_list.at(n_char).name;
-    QString str_show = char_list.at(n_char).name;
+    QString str_show = ao_app->get_showname(str_char);
```
Comment (Crystalwarrior) [inline src/courtroom.cpp:3166]: thx

**Review by Crystalwarrior — APPROVED** [review]
even tho I bring up a regression of certain oldschool behavior of double-clicking a category to stop a song, I honestly prefer this method of rightclicking to stop song lol

File: src/courtroom.cpp (hunk 7 above)
Comment (oldmud0) [reply src/courtroom.cpp:4096]: Over my dead body

Comment (Crystalwarrior) [reply src/courtroom.cpp:4096]: YES, DOWN WITH THE AWFUL UX

**Comment by oldmud0** [comment]
Once the conflicts are fixed I can merge

**Review by oldmud0 — APPROVED** [review]

**Review by Crystalwarrior — APPROVED** [review]
Ye

**Pull request merged**